supremm/summarize_mpi.py

```diff
@@ -215,37 +221,107 @@ def processjobs(config, opts, procid):
+                        numsent += 1
+                        logging.debug("Sent new job: %d sent, %d received", numsent, numreceived)
+                    else:
+                        # Initial batch
+                        comm.send(job, dest=numsent+1, tag=1)
+                        numsent += 1
+                        logging.debug("Initial Batch: %d sent, %d received", numsent, numreceived)
+
+                logging.debug("After all jobs sent: %d sent, %d received", numsent, numreceived)
+
+                # Get leftover results
+                while numsent > numreceived:
+                    comm.recv(source=MPI.ANY_SOURCE, tag=1)
+                    numreceived += 1
+                    logging.debug("Getting leftovers. %d sent, %d received", numsent, numreceived)
+
+                # Shut them down
+                for worker in xrange(numworkers):
+                    logging.debug("Shutting down: %d", worker+1)
+                    comm.send(None, dest=worker+1, tag=1)
+
+            # Worker
             else:
-                for job in dbif.get(None, None):
-                    summarizejob(job, config, resconf, plugins, preprocs, m, dbif, opts)
-
+                logging.debug("WORKER %d STARTING", procid)
+                while True:
+                    recvtries=0
+                    while not comm.Iprobe(source=0, tag=1):
+                        if recvtries < 1000:
+                            recvtries+=1
+                            continue
+                        # Sleep so we can instrument how efficient we are
+                        # Otherwise, workers spin on exit at the hidden mpi_finalize call.
+                        # If you care about maximum performance and don't care about wasted cycles, remove the Iprobe/sleep loop
+                        # Empirically, a tight loop with time.sleep(0.001) uses ~1% CPU
+                        time.sleep(0.001)
+                    job = comm.recv(source=0, tag=1)
+                    if job != None:
+                        logging.debug("Rank: %s, Starting: %s", procid, job.job_id)
+                        summarizejob(job, config, resconf, plugins, preprocs, m, dbif, opts)
+                        logging.debug("Rank: %s, Finished: %s", procid, job.job_id)
+                        comm.send(procid, dest=0, tag=1)
+                    else:
+                        # Got shutdown message
+                        break
 
 def main():
     """
     main entry point for script
     """
+
+    comm = MPI.COMM_WORLD
+
     opts = getoptions()
 
-    logging.basicConfig(format='%(asctime)s [%(levelname)s] %(message)s', datefmt='%Y-%m-%dT%H:%M:%S', level=opts['log'])
+    opts['threads'] = comm.Get_size()
+
+    logout = "mpiOutput-{}.log".format(comm.Get_rank())
+
+    # For MPI jobs, do something sane with logging.
+    setuplogger(logging.ERROR, logout, opts['log'])
+
     if sys.version.startswith("2.7"):
```

Comment:
The code that call captureWarnings() can be removed from here now that the setuplogger() function is being used to setup the logging.